ksqldb-engine/src/main/java/io/confluent/ksql/util/BinPackedStreamsMetadata.java

```diff
@@ -0,0 +1,199 @@
+import io.confluent.ksql.rest.entity.StreamsTaskMetadata;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.LagInfo;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.apache.kafka.streams.errors.StreamsUncaughtExceptionHandler;
+import org.apache.kafka.streams.state.StreamsMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class BinPackedStreamsMetadata {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BinPackedStreamsMetadata.class);
+
+  private final KafkaStreamsBuilder kafkaStreamsBuilder;
+  private QueryErrorClassifier errorClassifier ;
+  private KafkaStreams kafkaStreams;
+  private ImmutableMap<String, Object> streamsProperties;
+  private final QueryMetadataImpl.TimeBoundedQueue queryErrors;
+  private final Map<String, Topology> queryMap;
+  private final Map<QueryId, BinPackedPersistentQueryMetadataImpl> metadata;
+  private final Map<QueryId, QueryMetadata.Listener> listeners;
+
+  public BinPackedStreamsMetadata(final KafkaStreamsBuilder kafkaStreamsBuilder,
+                                  final int maxQueryErrorsQueueSize) {
+    this.kafkaStreamsBuilder = kafkaStreamsBuilder;
+    kafkaStreams = kafkaStreamsBuilder.build(null, streamsProperties);
+    queryErrors
+        = new QueryMetadataImpl.TimeBoundedQueue(Duration.ofHours(1), maxQueryErrorsQueueSize);
+    queryMap = new HashMap<>();
+    metadata = new HashMap<>();
+    listeners = new HashMap<>();
+  }
+
+  public void addQuery(final QueryErrorClassifier errorClassifier,
+                       final Map<String, Object> streamsProperties,
+                       final BinPackedPersistentQueryMetadataImpl binPackedPersistentQueryMetadata,
+                       final QueryId queryId) {
+    this.errorClassifier = errorClassifier;
+    this.streamsProperties =
+        ImmutableMap.copyOf(
+            Objects.requireNonNull(streamsProperties, "streamsProperties"));
+    this.metadata.put(queryId, binPackedPersistentQueryMetadata);
+    this.listeners.put(queryId, binPackedPersistentQueryMetadata.getListener());
+  }
+
+  protected StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse uncaughtHandler(
+      final Throwable e
+                                                                                         ) {
+    QueryError.Type errorType = QueryError.Type.UNKNOWN;
+    try {
+      errorType = errorClassifier.classify(e);
+    } catch (final Exception classificationException) {
+      LOG.error("Error classifying unhandled exception", classificationException);
+    } finally {
+      // If error classification throws then we consider the error to be an UNKNOWN error.
```

Comment:
A bit orthogonal to binpacking but more for error handling: when the thrown error is a runtime IllegalState/ArgumentException, what error type do we classify to? And do we still handle as `REPLACE_THREAD`?

Reply:
Right now we always replace the thread. With bin packing we might able to remove the problematic topology on non recoverable errors once we have classification on that axis

Reply:
Sounds good.

We had some discussions on AK regarding these types of exceptions, since they indicate some bugs in the KS code. In the future maybe we can improve KS to not pass this to exceptional handler at all and just fail-stop immediately, but for now we'd need KSQL layer to handle them as fatal.